Fix MemoryStore#write with unless_exist and namespace

sbfaulkner opened the pull request:
### Motivation / Background

This Pull Request has been created to fix `ActiveSupport::Cache::MemoryStore#write` support for `unless_exist: true` when using the `namespace` option as well.

### Detail

This Pull Request fixes `MemoryStore#write_entry` so that it passes a nil namespace to `exist?`. This is necessary because `exist?` since the _key_ has already been normalized to include the namespace.

### Additional information

In addition to adding a test for `write` with `unless_exist: true` using a `namespace` as well, the existing tests for `write` with `unless_exist: true` were moved to `MemoryStoreTest` (from `MemoryStorePruningTest`).

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.

jonathanhefner commented on activesupport/lib/active_support/cache/memory_store.rb:
```diff
@@ -210,7 +210,10 @@ def read_entry(key, **options)
           payload = serialize_entry(entry, **options)
           synchronize do
-            return false if options[:unless_exist] && exist?(key)
+            if options[:unless_exist]
+              entry = read_entry(key, **options)
+              return false if entry && !entry.expired? && !entry.mismatched?(normalize_version(key, options))
+            end
```
So the problem with using `exist?(key)` is that, by this point, `key` is already namespaced and then [`exist?` _also_ applies the namespace from the base options](https://github.com/rails/rails/blob/139c5678aa5d9a4209ca06919abea7a9da449b36/activesupport/lib/active_support/cache.rb#L711), correct?

In that case, we could just specify `namespace: nil` instead of duplicating `exist?`'s logic:

```suggestion
            return false if options[:unless_exist] && exist?(key, namespace: nil)
```

sbfaulkner replied:
calling exist could bring other side-effects too -- eg. it calls merge_options (again - it's already been done further up the current call stack)

jonathanhefner commented on activesupport/test/cache/behaviors/cache_logging_behavior.rb:
```diff
@@ -77,7 +86,7 @@ def assert_logs(pattern, &block)
       assert_match pattern, io.string
     end
 
-    def key_pattern(key, namespace: @namespace)
+    def key_pattern(key, namespace:)
```
Is this change and all the above changes to accommodate setting `@namespace` in `MemoryStoreTest`?  Or is there some other benefit?

sbfaulkner replied:
I do understand though that testing an un-namespaced memcache or redis store is problematic, so will change this back

jonathanhefner commented on activesupport/test/cache/stores/memory_store_test.rb:
```diff
@@ -7,6 +7,8 @@
 class MemoryStoreTest < ActiveSupport::TestCase
   def setup
     @cache = lookup_store(expires_in: 60)
+    @namespace = "test-#{Random.rand(16**32).to_s(16)}"
+    @namespaced_cache = lookup_store(expires_in: 60, namespace: @namespace)
```
The Redis and Memcache tests use `@namespace` to prevent tests from interfering with each other.  That isn't a concern for `MemoryStore`, so I think we can omit these instance variables, especially since they are only used for a single test.  (See my suggestion further below.)

jonathanhefner commented on activesupport/test/cache/stores/memory_store_test.rb:
```diff
@@ -61,6 +63,26 @@ def test_nil_coder_bypasses_mutation_safeguard
+  end
+
+  def test_namespaced_write_with_unless_exist
+    assert_equal true, @namespaced_cache.write(1, "aaaaaaaaaa")
+    assert_equal false, @namespaced_cache.write(1, "aaaaaaaaaa", unless_exist: true)
+    @namespaced_cache.write(1, nil)
+    assert_equal false, @namespaced_cache.write(1, "aaaaaaaaaa", unless_exist: true)
```
```suggestion
    @cache = lookup_store(expires_in: 60, namespace: "foo")
    
    assert_equal true, @cache.write(1, "aaaaaaaaaa")
    assert_equal false, @cache.write(1, "aaaaaaaaaa", unless_exist: true)
    @cache.write(1, nil)
    assert_equal false, @cache.write(1, "aaaaaaaaaa", unless_exist: true)
```

sbfaulkner replied:
once I get the other PRs merged, I think this will become more necessary, but 🤷🏻

sbfaulkner committed:
Fix MemoryStore#write with unless_exist and namespace

Updates `MemoryStore#write_entry` to pass a `nil` `namespace` to
`exist?`, which expects a _name_ rather than a an already "normalized"
_key_. This fixes a bug where `unless_exist` would overwrite any
existing entry if a `namespace` was used.

sbfaulkner commented:
@jonathanhefner I've implemented your suggestions.

jonathanhefner commented on activesupport/test/cache/behaviors/cache_logging_behavior.rb:
```diff
@@ -56,18 +56,27 @@ def test_exist_logging
     assert_logs("Cache read_multi: 1 key(s)") { @cache.read_multi("foo") }
     assert_logs("Cache read_multi: 2 key(s)") { @cache.read_multi("foo", "bar") }
+
+    assert_logs("Cache read_multi: 1 key(s)") { @cache.read_multi("foo", namespace: "bar") }
+    assert_logs("Cache read_multi: 2 key(s)") { @cache.read_multi("foo", "bar", namespace: "bar") }
```
Specifying a static namespace could cause tests to interfere with each other in the same way as having no namespace.

We could instead generate a random namespace for these assertions, but I'm not sure how much that buys us because (1) the expected value for these assertions is the same with or without a namespace, and (2) the Redis and Memcache tests are implicitly testing this code with a namespace.

I am more inclined to simply omit these assertions.  Even more so since they are not related to the fix.

sbfaulkner replied:
I see what you mean and will consider an alternative - my change is maybe more applicable for #50200 which involves read_multi, but still needs more thought re: use of namespace

sbfaulkner replied:
omitting as suggested

sbfaulkner commented:
@jonathanhefner updated again per suggestions

jonathanhefner commented:
Thank you, @sbfaulkner! :tada:
